langchain/src/util/googlevertexai-connection.ts

```diff
@@ -106,29 +191,229 @@ export class GoogleVertexAILLMConnection<
+
+    // Loop while we don't have something to return AND we have something in the buffer
+    let index = 0;
+    while (ret === null && this._buffer.length > index) {
+      // Advance to the next close bracket after our current index
+      index = this._buffer.indexOf("}", index + 1);
+
+      // If we don't find one, exit with null
+      if (index === -1) {
+        return null;
+      }
+
+      // If we have one, try to turn it into an object to return
+      try {
+        const objStr = this._buffer.substring(0, index + 1);
+        ret = JSON.parse(objStr);
+
+        // We only get here if it parsed it ok
+        // If we did turn it into an object, remove it from the buffer
+        this._buffer = this._buffer.slice(index + 1);
+      } catch (xx) {
+        // It didn't parse it correctly, so we swallow the exception and continue
+      }
+    }
+
+    return ret;
+  }
+
+  _simplifyObject(obj: unknown): object {
+    return simpleValue(obj) as object;
+  }
+
+  // Set up a potential Promise that the handler can resolve.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  _chunkResolution: (chunk: any) => void;
+
+  // If there is no Promise (it is null), the handler must add it to the queue
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  _chunkPending: Promise<any> | null = null;
+
+  // A queue that will collect chunks while there is no Promise
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  _chunkQueue: any[] = [];
+
+  /**
+   * Register that we have another chunk available for consumption.
+   * If we are waiting for a chunk, resolve the promise waiting for it immediately.
+   * If not, then add it to the queue.
+   * @param chunk
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  _handleChunk(chunk: any): void {
+    if (this._chunkPending) {
+      this._chunkResolution(chunk);
+      this._chunkPending = null;
+    } else {
+      this._chunkQueue.push(chunk);
+    }
+  }
+
+  /**
+   * Get the next chunk that is coming from the stream.
+   * This chunk may be null, usually indicating the last chunk in the stream.
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  async nextChunk(): Promise<any> {
+    if (this._chunkQueue.length > 0) {
+      // If there is data in the queue, return the next queue chunk
+      return this._chunkQueue.shift() as GenerationChunk;
+    } else {
+      // Otherwise, set up a promise that handleChunk will cause to be resolved
+      this._chunkPending = new Promise((resolve) => {
+        this._chunkResolution = resolve;
+      });
+      return await this._chunkPending;
+    }
+  }
+
+  /**
+   * Is the stream done?
+   * A stream is only done if all of the following are true:
+   * - There is no more data to be added to the text buffer
+   * - There is no more data in the text buffer
+   * - There are no chunks that are waiting to be consumed
+   */
+  get streamDone(): boolean {
```

Comment:
Would be awesome to implement as an iterator - I can take a crack at it later this week